### ISSUE TITLE
How to contribute

### ISSUE THREAD

**Issue by rdunklau** [lelit/pglast#13]
Hello,

I'm currently using your library to build a tool, and running it on our pg_dumps and migration script made me implement / override printers for a large amount of statement types. I did not bother with handling indentation for now as that's not my intended use, but I managed to get no errors on our 300k+ lines of SQL code when running prettify with the safety belt on. (there's still a lot to do if we want to handle everything present the PostgreSQL regression tests though :) )

I have implemented support for things we use, which may be incomplete, for the following nodes:

 ```
AccessPriv
AlterDatabaseSetStmt
AlterDefaultPrivilegesStmt
AlterEnumStmt
AlterFunctionStmt
AlterObjectSchemaStmt
AlterOwnerStmt
AlterRoleStmt
AlterSeqStmt
AlterTableCmd
AlterTableStmt
ClusterStmt
CompositeTypeStmt
CreateEnumStmt
CreateFunctionStmt
CreatePolicyStmt
CreateTrigStmt
DiscardStmt
DoStmt
GrantRoleStmt
GrantStmt
GroupingSet
LockStmt
RenameStmt
TriggerTransition
VacuumStmt
VariableSetStmt
ViewStmt
```

In addition, I've overridden printers for those nodes (adding support for unsupported options / handling corner cases):

```
A_ArrayExpr
A_Expr
ColumnDef
Constraint
CreateTableAsStmt
DefineStmt
DropStmt
IndexElem
NullTest
ObjectWithArgs
RowExpr
TypeCast
```

I obviously want to contribute that back, but I don't really know how to do that. Would you like indivual PRs for every node or a huge one for the whole lot  ? How should that code be tested ?

Thank you in advance for your help regarding this.

**Comment by lelit** [comment]
That's great, thank you!

I know that lots of nodes are missing, but given the impressive number of them, I think the most viable way is to add them as the need arise... I could not find a comprehensive collection of all possible statements. BTW, how are the "regression tests" you mention written? Could you point me there?

Wrt contributing back, I'm fine with a single PR, as long as there are tests against all added nodes, keeping test coverage as close to 100% as possible. That basically means adding SQL statements to `tests/test_d?l_printers.py` (for example, in `tests/test_dml_printers.py` there's a `SELECTS` multiline string, with SQL statements separated by an empty line...) asserting that all the nodes get exercized.

Cannot promise on when I can do that, but I'm willing to take care of the _prettification_ of the new nodes.

About the "printer overrides", that should be properly integrated, if possible.

Let me know if you need further advice!

**Comment by rdunklau** [comment]
Thank you, I will try to do that as soon as possible, hopefully this week.

The regression tests for PostgreSQL are mostly here: https://git.postgresql.org/gitweb/?p=postgresql.git;a=tree;f=src/test/regress/sql;h=381270746f8f42574dedd946c190e6a1be89d4ee;hb=HEAD

They consist in a set of sql file, each associated with an expected output. I think this makes for a great reference of which features are supported or not, since they should exercise every possible statement (in theory, I don't know the exact coverage though).

**Comment by apnewberry** [comment]
> I could not find a comprehensive collection of all possible statements. 

This isn't specific to postgres, but it might be useful: [BNF Grammars for SQL-92, SQL-99 and SQL-2003](https://ronsavage.github.io/SQL/)

**Comment by lelit** [comment]
This has been released as of version 1.3. Thank you!
